### PR TITLE
DRILL-4476: Allow UnionAllRecordBatch to manager situations where lef…

### PR DESCRIPTION
**Pull request by hsuanyi**
…t input side or both sides come(s) from empty source(s).

### CODE REVIEW

**Comment by hsuanyi** [comment]
@amansinha100 can you help review this?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java [hunk 1]
```diff
@@ -162,6 +162,25 @@ private IterOutcome doWork() throws ClassTransformationException, IOException, S
     allocationVectors = Lists.newArrayList();
     transfers.clear();
 
+    // If both sides of Union-All are empty
+    if(unionAllInput.isBothSideEmpty()) {
+      for(int i = 0; i < outputFields.size(); ++i) {
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: If both sides are empty, wouldn't outputFields.size() be 0 ?  Or is outputFields populated from a previous batch, which implies both sides may not have been empty.  Can you clarify ?

File: exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java [hunk 2]
```diff
@@ -527,6 +532,52 @@ public void testUnionAllRightEmptyJson() throws Exception {
       .build().run();
   }
 
+  @Test
+  public void testUnionAllLeftEmptyJson() throws Exception {
+    final String rootEmpty = FileUtils.getResourceAsFile("/project/pushdown/empty.json").toURI().toString();
+    final String rootSimple = FileUtils.getResourceAsFile("/store/json/booleanData.json").toURI().toString();
+
+    final String queryRightEmpty = String.format(
```
Comment (amansinha100) [inline exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java]: The name suggests right is empty but the query has left empty...

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java [hunk 3]
```diff
@@ -491,6 +556,25 @@ private void inferOutputFieldsFromLeftSide() {
       }
     }
 
+    private void inferOutputFieldsFromRightSide() {
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: It seems like the inferencing is repeated a few times.  You could have a single function like inferOutputFieldNameAndType(outputFieldNames,  batch).  The field type is always inferred from the supplied batch.  For the output names, if the outputFieldNames is null, then use the output field names as well as type from that batch (this would be the case for left input).  If not null, then use the field names.  Would that work ?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java [hunk 4]
```diff
@@ -294,13 +313,41 @@ public UnionAllInput(UnionAllRecordBatch unionAllRecordBatch, RecordBatch left,
...
+            whileLoop:
+            while(leftSide.getRecordBatch().getRecordCount() == 0) {
+              iterLeft = leftSide.nextBatch();
+
+              switch(iterLeft) {
+                case STOP:
+                case OUT_OF_MEMORY:
+                  return iterLeft;
+
+                case NONE:
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: Is the NONE state not reached outside this while loop ? It wasn't clear why we need the inner while loop for the left input considering that you don't do that for the right input.

File: exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java [hunk 5]
```diff
@@ -527,6 +532,52 @@ public void testUnionAllRightEmptyJson() throws Exception {
       .build().run();
   }
 
+  @Test
+  public void testUnionAllLeftEmptyJson() throws Exception {
```
Comment (amansinha100) [inline exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java]: Are there any union-all tests with  WHERE 1 = 0 filters on the left and/or right inputs ?  If not, could you add such tests ?  We should distinguish between a truly empty input vs. one that occurs as a result of a False filter.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java (hunk 1 above)
Comment (hsuanyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: For example, you could do:
select a, b, c from empty union all select a, b, c from empty
Then, the output should be (a, b, c) as column names.

File: exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java (hunk 2 above)
Comment (hsuanyi) [inline exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java]: addressed

File: exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java (hunk 5 above)
Comment (hsuanyi) [inline exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java]: Added a few for both union-all and union-distinct

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java (hunk 4 above)
Comment (hsuanyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: The while is for the case where the first few batches are empty and then union-all receive NONE.

So we need to replicate this logic to right side too.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java (hunk 3 above)
Comment (hsuanyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: Inferencing happens only once when Drill receives the very first batches from left and right. (Schema-change is not yet supported for Union-All).

Let me summarize the inference in four different situations:
First of all, the field names are always determined by the left side (even when the left side is from empty file, we have the column names. Please see the comment above.)

1. Left: non-empty; Right: non-empty=> types determined by both sides with implicit casting involved
2. Left: empty; Right: non-empty=> type from the right
3. Left: non-empty; Right: empty=> types from the left
4. Left: empty; Right: empty=> types are nullable integer

**Comment by hsuanyi** [comment]
@amansinha100 I consolidated the two logic into one:
https://github.com/hsuanyi/incubator-drill/blob/DRILL-4476/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java#L597

Can you help review again? Thanks.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java (hunk 1 above)
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: Ok, I suppose we are not supporting the SELECT * case, so the output fields only depends on what is specified in the SELECT list, not from the table.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java (hunk 4 above)
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: Could you add comments here about why we need to keep reading after getting OK_NEW_SCHEMA.  Also, do any of the union-all tests (either unit or functional) exercise this code ? i.e where the left input has 2 or more empty files ?

**Comment by amansinha100** [comment]
Updated changes LGTM.  +1

**Comment by hsuanyi** [comment]
Comments were added to the code; Also passed all the pre commit test. Can you help commit this patch ?
